rust_src/src/eval.rs

```diff
@@ -97,4 +106,84 @@ pub fn progn(body: LispObject) -> LispObject {
         .map_or_else(|| LispObject::constant_nil(), LispObject::from_raw)
 }
 
+/// Evaluate BODY sequentially, discarding its value.
+#[no_mangle]
+pub extern "C" fn prog_ignore(body: Lisp_Object) {
+    progn(LispObject::from_raw(body));
+}
+
+/// Eval FIRST and BODY sequentially; return value from FIRST.
+/// The value of FIRST is saved during the evaluation of the remaining args,
+/// whose values are discarded.
+/// usage: (prog1 FIRST BODY...)
+#[lisp_fn(min = "1", unevalled = "true")]
+pub fn prog1(args: LispObject) -> LispObject {
+    let (first, body) = args.as_cons_or_error().as_tuple();
+
+    let val = unsafe { eval_sub(first.to_raw()) };
+    progn(body);
+    LispObject::from_raw(val)
+}
+
+/// Eval FORM1, FORM2 and BODY sequentially; return value from FORM2.
+/// The value of FORM2 is saved during the evaluation of the
+/// remaining args, whose values are discarded.
+/// usage: (prog2 FORM1 FORM2 BODY...)
+#[lisp_fn(min = "2", unevalled = "true")]
+pub fn prog2(args: LispObject) -> LispObject {
+    let (form1, tail) = args.as_cons_or_error().as_tuple();
+
+    unsafe { eval_sub(form1.to_raw()) };
+    prog1(tail)
+}
+
+/// Set each SYM to the value of its VAL.
+/// The symbols SYM are variables; they are literal (not evaluated).
+/// The values VAL are expressions; they are evaluated.
+/// Thus, (setq x (1+ y)) sets `x' to the value of `(1+ y)'.
+/// The second VAL is not computed until after the first SYM is set, and so on;
+/// each VAL can use the new value of variables set earlier in the `setq'.
+/// The return value of the `setq' form is the value of the last VAL.
+/// usage: (setq [SYM VAL]...)
+#[lisp_fn(min = "0", unevalled = "true")]
+pub fn setq(args: LispObject) -> LispObject {
+    let mut val = args;
+
+    let mut it = args.iter_cars().enumerate();
+    while let Some((nargs, sym)) = it.next() {
+        let (_, arg) = it.next().unwrap_or_else(|| {
+            xsignal!(
+                Qwrong_number_of_arguments,
+                LispObject::from_raw(Qsetq),
+                LispObject::from(nargs + 1)
+            );
+        });
+
+        val = LispObject::from_raw(unsafe { eval_sub(arg.to_raw()) });
+
+        let mut lexical = false;
+
+        // Like for eval_sub, we do not check declared_special here since
+        // it's been done when let-binding.
+        // N.B. the check against nil is a mere optimization!
+        if unsafe { globals.f_Vinternal_interpreter_environment != Qnil } && sym.is_symbol() {
+            let binding = assq(
+                sym,
+                LispObject::from_raw(unsafe { globals.f_Vinternal_interpreter_environment }),
+            );
+            if let Some(binding) = binding.as_cons() {
```

Comment:
The C code only checks that binding is non-nil here. However, since we call .set_cdr afterwards, I think checking that it's a cons is more robust :)